\documentclass[a4paper, 9pt]{article}

<<style-Sweave, eval=TRUE, echo=FALSE, results=tex>>=
BiocStyle::latex()
@

\usepackage{hyperref}
\usepackage{amsmath, amsthm, amssymb}
\usepackage{xfrac}

\usepackage{fullpage}

\usepackage{marginnote}
\usepackage{graphicx}

\usepackage[table]{xcolor} %http://ctan.org/pkg/xcolor

\usepackage[numbers]{natbib}
\usepackage{algorithmic}
\usepackage{algorithm}

\usepackage{url}

\usepackage{placeins}


\usepackage{xspace}

\newcommand{\CAPRESE}{\textsc{caprese}}
\newcommand{\TRONCO}{\textsc{tronco}}

\usepackage{fullpage}

\begin{document}


\title{Using the \TRONCO{} package}


\author{
Marco Antoniotti\footnote{Dipartimento di Informatica Sistemistica e Comunicazione, Universit\'a degli Studi Milano-Bicocca
Milano, Italy.} \and
Giulio Caravagna$^\ast$ \and
Luca De Sano$^\ast$ \and
Alex Graudenzi$^\ast$ \and
Ilya Korsunsky\footnote{Courant Institute of Mathematical Sciences, New York University, New York, USA.} \and
Mattia Longoni$^\ast$ \and
Loes Olde Loohuis\footnote{Center for Neurobehavioral Genetics, University of California, Los Angeles, USA.} \and
Giancarlo Mauri$^\ast$ \and
Bud Mishra$^\dagger$ \and
Daniele Ramazzotti$^\ast$ 
}

\date{\today}
\maketitle


\begin{center}
\begin{minipage}[h]{0.75\textwidth}
\textbf{Abstract.} Genotype-level {\em cancer progression models} describe the ordering of accumulating mutations, e.g.,  somatic mutations / copy number variations, during  cancer development. These graphical models help understand the ``causal structure'' involving events promoting cancer progression, possibly predicting complex patterns characterising  genomic  progression of a cancer. Reconstructed models can be used to better characterise genotype-phenotype relation, and suggest novel targets for therapy design.

\TRONCO{} ({\sc tr}{\em anslational} {\sc onco}{\em logy}) is a \textsc{r} package  aimed at collecting  state-of-the-art algorithms to infer
\emph{progression models}  from \emph{cross-sectional} data, i.e., data collected from independent patients which does not necessarily incorporate any evident temporal information. These algorithms require a binary input matrix where: $(i)$ each row represents a patient genome, $(ii)$ each column an event relevant to the progression (a priori selected) and a $0/1$ value models the absence/presence of a certain mutation in a certain patient.

 

The current first version of \TRONCO{}
implements the \CAPRESE{} algorithm ({\sc ca}{\em ncer} {\sc pr}{\em ogression} {\sc e}{\em xtraction} {\em with} {\sc s}{\em ingle} {\sc e}{\em dges}) to infer possible progression models arranged as \emph{trees};
cfr.
\begin{itemize}
\item \emph{Inferring tree causal models of cancer progression with
    probability raising},  L. Olde Loohuis, G. Caravagna,
  A. Graudenzi, D. Ramazzotti, G. Mauri,  M. Antoniotti and
  B. Mishra. {PLoS One}, \emph{to appear}.
\end{itemize}
This vignette shows how to  use  \TRONCO{} to infer a tree model of
ovarian cancer progression from CGH data of copy number alterations (classified as gains or losses over chromosome's arms). The dataset used is
available in the  SKY/M-FISH database.
The reference manual for \TRONCO{} is available in the package.

\end{minipage}
\end{center}

\vspace{1.0cm}


\SweaveOpts{concordance=TRUE}

\paragraph{\large Requirements: } You must  have \texttt{rgraphviz}
installed to  use the package, see \texttt{Bioconductor.org}.


\paragraph{\large Event selection}{\ }\\

First, load \TRONCO{} in your \textsc{R} console and the example \textit{"dataset"}. 
<<>>=
library(TRONCO)
data(aCML)
hide.progress.bar <<- TRUE
@

\paragraph{ We use \texttt{show} function to get a short summary of the aCML dataset }
<<>>=
show(data)
@

\paragraph{ These are all the events it contains }
<<>>=
as.events(data)
@

\paragraph{ Which account for alterations in the following genes }
<<>>=
as.genes(data)
@

\paragraph{ These are \texttt{SETBP1} alterations across input samples }
<<>>=
as.gene(data, genes='SETBP1')
@

\paragraph{ These are the genes for which we found a literature supporting the patterns
that we include below. References are in the main \textit{CAPRI} paper. }
<<>>=
gene.hypotheses = c('KRAS', 'NRAS', 'IDH1', 'IDH2', 'TET2', 'SF3B1', 'ASXL1')
@

\paragraph{ Regardless the distinct types of mutations that we included, we want to
select only genes altered in $5\%$ of the cases. Thus we first transform
data in \textit{"Alteration"} (collapsing all event types for the same gene), and
then we use select only those events }
<<>>=
alterations = events.selection(as.alterations(data), filter.freq = .05)
@

\paragraph{ We visualize the selected genes. This plot has no title since name annotation
is not copied by \texttt{events.selection} }
<<onco1, fig=TRUE,include=TRUE>>=
dummy = oncoprint(alterations)
@
\incfig[ht]{vignette-onco1}{0.9\textwidth}{Oncoprint output}{}

\FloatBarrier

\paragraph{\large Adding Hypotheses}{\ }\\

\paragraph{ Then to reconstruct the aCML model we select from \texttt{data} which have been
selected in \texttt{alteration} - via \texttt{as.genes(alterations)} or that are
part of the prior in \texttt{gene.hypotheses}. We use \texttt{filter.in.names} to
force selection of all the events involving those genes from \texttt{data} }
<<>>=
hypo = events.selection(data, filter.in.names=c(as.genes(alterations), gene.hypotheses))
hypo = annotate.description(hypo, 'CAPRI - Bionformatics aCML data (selected events)')
@

\paragraph{ We show selected data and we annotate genes in \texttt{gene.hypotheses} to identify
them. Samples names are also shown }
<<onco2, fig=TRUE,include=FALSE>>=
dummy = oncoprint(hypo,  gene.annot = list(priors= gene.hypotheses), sample.id = T)
@
\incfig[ht]{vignette-onco2}{0.9\textwidth}{Oncoprint output}{}

\FloatBarrier

\paragraph{ We now add the hypotheses that are described in CAPRI's manuscript }

\paragraph{ Add hypotheses of hard exclusivity (XOR) for NRAS/KRAS events (Mutation). The hypothesis is tested
against all other dataset events }
<<>>=
hypo = hypothesis.add(hypo, 'NRAS xor KRAS', XOR('NRAS', 'KRAS'))
@

\paragraph{ Here we try to include also a soft exclusivity (OR) pattern but, since its \textit{"signature"}
is the same of the hard one, it will not be included. The code below is commented because it gives errors. }
<<>>=
### do not run
# hypo = hypothesis.add(hypo, 'NRAS or KRAS',  OR('NRAS', 'KRAS'))
###
@

\paragraph{ For the sake to better highlight the perfect (hard) exclusivity between NRAS/KRAS 
mutations one can visualize their alterations }
<<onco3, fig=TRUE,include=FALSE>>=
dummy = oncoprint(events.selection(hypo, filter.in.names = c('KRAS', 'NRAS')))
@
\incfig[ht]{vignette-onco3}{0.9\textwidth}{Oncoprint output}{}

\FloatBarrier

\paragraph{ This is as above, but includes other events. Again, we can include only the hard exclusivity pattern }
<<>>=
hypo = hypothesis.add(hypo, 'SF3B1 xor ASXL1', XOR('SF3B1', OR('ASXL1')), '*')
### do not run
# hypo = hypothesis.add(hypo, 'SF3B1 or ASXL1', OR('SF3B1', OR('ASXL1')), '*')
###
@

\paragraph{ We now do the same for TET2 and IDH2. In this case 3 events for TET2 are present, which are
\textit{"Ins/Del"}, \textit{"Missense point"} and \textit{"Nonsense point"}. For this reason, since we are not specifying
a subset of such events all TET2 alterations are used. Since these show a perfect hard exclusivity
trend these will be included in XOR. }
<<>>=
as.events(hypo, genes = 'TET2') 
hypo = hypothesis.add(hypo, 'TET2 xor IDH2', XOR('TET2', 'IDH2'), '*')
### do not run
# hypo = hypothesis.add(hypo,  'TET2 or IDH2', OR('TET2', 'IDH2'), '*'))
###
@

<<onco4, fig=TRUE,include=FALSE>>=
dummy = oncoprint(events.selection(hypo, filter.in.names = c('TET2', 'IDH2')))
@
\incfig[ht]{vignette-onco4}{0.9\textwidth}{Oncoprint output}{}

\FloatBarrier

\paragraph{ For every gene that has more than one event associated we also add a soft exclusivity pattern
for its events }
<<>>=
hypo = hypothesis.add.homologous(hypo)
@

\paragraph{ The dataset input to CAPRI is shown }
<<onco5, fig=TRUE,include=FALSE>>=
dummy = oncoprint(hypo,  gene.annot = list(priors= gene.hypotheses), sample.id = T)
@
\incfig[ht]{vignette-onco5}{0.9\textwidth}{Oncoprint output}{}

\FloatBarrier

\paragraph{\large Model reconstruction}{\ }\\

\paragraph{ We execute CAPRI with its default parameter: we use both AIC/BIC regularizators, Hill-climbing
exhaustive bootstrap (100 replicates for Wilcoxon testing), p-value 0.05 and we set seed }
<<>>=
model = tronco.capri(hypo, boot.seed = 12345, regularization='bic')
@

\paragraph{ We can plot the reconstructed model. We set some parameters to get a fancy plot; confidence
is shown as temporal priority and probability raising (selective advantage scores) and 
hypergeometric testing (goodness of input data). }
<<figplot, fig=TRUE,include=FALSE>>=
tronco.plot(model,  
  fontsize = 13,  
  scale.nodes = .6,   
  confidence = c('tp', 'pr', 'hg'),   
  height.logic = 0.25,  
  legend.cex = .5, 
  pathways =  list(priors= gene.hypotheses))
@
\incfig[ht]{vignette-figplot}{0.9\textwidth}{aCML Reconstructed model}
{Pre bootstrap.}

\FloatBarrier

\paragraph{\large Bootstrapping data}{\ }\\

<<>>=
model.boot = tronco.bootstrap(model, nboot=6)
@

<<figplotboot, fig=TRUE,include=FALSE>>=
tronco.plot(model.boot,
            fontsize = 13,  
            scale.nodes = .6,
            confidence=c('npb'),
            height.logic = 0.25,  
            legend.cex = .5)
@
\incfig[ht]{vignette-figplotboot}{0.9\textwidth}{aCML Reconstructed model}
{After bootstrap.}


\end{document}



% buggone -> fix http://stackoverflow.com/questions/12481267/in-r-how-to-prevent-blank-page-in-pdf-when-using-gridbase-to-embed-subplot-insi ??

