\documentclass[a4paper, 9pt]{article}

<<style-Sweave, eval=TRUE, echo=FALSE, results=tex>>=
BiocStyle::latex()
@

% \VignetteIndexEntry{An R Package for TRanslational ONCOlogy}

\usepackage{hyperref}
\usepackage{amsmath, amsthm, amssymb}
\usepackage{xfrac}

\usepackage{fullpage}

\usepackage{marginnote}
\usepackage{graphicx}

\usepackage[table]{xcolor} %http://ctan.org/pkg/xcolor

\usepackage[numbers]{natbib}
\usepackage{algorithmic}
\usepackage{algorithm}

\usepackage{url}

\usepackage{placeins}


\usepackage{xspace}

\newcommand{\CAPRESE}{\textsc{caprese}}
\newcommand{\TRONCO}{\textsc{tronco}}

\usepackage{fullpage}

\begin{document}


\title{Using the \TRONCO{} package}


\author{
Marco Antoniotti\footnote{Dipartimento di Informatica Sistemistica e Comunicazione, Universit\'a degli Studi Milano-Bicocca
Milano, Italy.} \and
Giulio Caravagna$^\ast$ \and
Luca De Sano$^\ast$ \and
Alex Graudenzi$^\ast$ \and
Ilya Korsunsky\footnote{Courant Institute of Mathematical Sciences, New York University, New York, USA.} \and
Mattia Longoni$^\ast$ \and
Loes Olde Loohuis\footnote{Center for Neurobehavioral Genetics, University of California, Los Angeles, USA.} \and
Giancarlo Mauri$^\ast$ \and
Bud Mishra$^\dagger$ \and
Daniele Ramazzotti$^\ast$ 
}

\date{\today}
\maketitle


\begin{center}
\begin{minipage}[h]{0.75\textwidth}
\textbf{Abstract.} Genotype-level {\em cancer progression models} describe the temporal ordering in which genomic alterations such as somatic mutations and copy number alterations tend to fixate and accumulate during cancer formation and progression. These graphical models can describe trends of \textit{natural selection} across a population of independent tumour samples (cross-sectional data), or reconstruct the clonal evolution in a single patient's tumour (multi-region or single-cell data). In terms of application, such models can be used to better elucidate genotype-phenotype relation, predict cancer hallmarks and outcome of personalised treatment as well as suggest novel targets for therapy design.
\\

\TRONCO{} ({\sc tr}{\em anslational} {\sc onco}{\em logy}) is a \textsc{R} package which collects 
algorithms to infer progression models from Bernoulli 0/1 profiles of genomic
alterations across a tumor sample. Such profiles are usually visualised as a 
binary input matrix where each row represents a patient's sample (e.g., the 
result of a sequenced tumor biopsy), and each column an event relevant to the 
progression (a certain type of somatic mutation, a focal or higher-level 
chromosomal copy number alteration etc.); a 0/1 value models the absence/presence
of that alteration in the sample. In this version of TRONCO such profiles can 
be readily imported by boolean matrices and MAF/GISTIC files. The package provides
various functions to editing, visualise and subset such data, as well as functions 
to query the Cbio portal for cancer genomics. This version of TRONCO comes with
the parallel implementations the CAPRESE  [PLoS ONE 9(12): e115570] and CAPRI 
[Bioinformatics, doi:10.1093/bioinformatics/btv296] algorithms to infer possible
progression models arranged as trees, or general direct acyclic graphs. 
Bootstrap functions to assess the parametric, non-prametric and statistical 
confidence of every inferred model are also provided. The package comes with 
some data available as well, which include the dataset of Atypical Chronic Myeloid 
Leukemia samples provided by Piazza et al., Nat. Genet., 45 (2013), and examples.

\end{minipage}
\end{center}

\vspace{1.0cm}


\SweaveOpts{concordance=TRUE}

\paragraph{\large Requirements: } You must  have \texttt{rgraphviz}
installed to  use the package, see \texttt{Bioconductor.org}.


\paragraph{\large Event selection}{\ }\\

First, load \TRONCO{} in your \textsc{R} console and the example \textit{"dataset"}. 
<<>>=
library(TRONCO)
data(aCML)
hide.progress.bar <<- TRUE
@

\paragraph{ We use \texttt{show} function to get a short summary of the aCML dataset }
<<>>=
show(aCML)
@

\paragraph{ These are all the events it contains }
<<>>=
as.events(aCML)
@

\paragraph{ Which account for alterations in the following genes }
<<>>=
as.genes(aCML)
@

\paragraph{ These are \texttt{SETBP1} alterations across input samples }
<<>>=
as.gene(aCML, genes='SETBP1')
@

\paragraph{ These are the genes for which we found a literature supporting the patterns
that we include below. References are in the main \textit{CAPRI} paper. }
<<>>=
gene.hypotheses = c('KRAS', 'NRAS', 'IDH1', 'IDH2', 'TET2', 'SF3B1', 'ASXL1')
@

\paragraph{ Regardless the distinct types of mutations that we included, we want to
select only genes altered in $5\%$ of the cases. Thus we first transform
data in \textit{"Alteration"} (collapsing all event types for the same gene), and
then we use select only those events }
<<>>=
alterations = events.selection(as.alterations(aCML), filter.freq = .05)
@

\paragraph{ We visualize the selected genes. This plot has no title since name annotation
is not copied by \texttt{events.selection} }
<<>>=
dummy = oncoprint(alterations)
@
<<include=FALSE,echo=FALSE>>=
capture.output(oncoprint(alterations, file='onco-1.pdf'), file='NUL')
@
\incfig[ht]{onco-1}{0.9\textwidth}{Oncoprint output}{}

\FloatBarrier

\paragraph{\large Adding Hypotheses}{\ }\\

\paragraph{ Then to reconstruct the aCML model we select from \texttt{data} which have been
selected in \texttt{alteration} - via \texttt{as.genes(alterations)} or that are
part of the prior in \texttt{gene.hypotheses}. We use \texttt{filter.in.names} to
force selection of all the events involving those genes from \texttt{data} }
<<>>=
hypo = events.selection(aCML, filter.in.names=c(as.genes(alterations), gene.hypotheses))
hypo = annotate.description(hypo, 'CAPRI - Bionformatics aCML data (selected events)')
@

\paragraph{ We show selected data and we annotate genes in \texttt{gene.hypotheses} to identify
them. Samples names are also shown }
<<>>=
dummy = oncoprint(hypo,  gene.annot = list(priors= gene.hypotheses), sample.id = T)
@
<<include=FALSE,echo=FALSE>>=
capture.output(oncoprint(hypo,  gene.annot = list(priors= gene.hypotheses), sample.id = T, file='onco-2.pdf'), file='NUL')
@
\incfig[ht]{onco-2}{0.9\textwidth}{Oncoprint output}{}

\FloatBarrier

\paragraph{ We now add the hypotheses that are described in CAPRI's manuscript }

\paragraph{ Add hypotheses of hard exclusivity (XOR) for NRAS/KRAS events (Mutation). The hypothesis is tested
against all other dataset events }
<<>>=
hypo = hypothesis.add(hypo, 'NRAS xor KRAS', XOR('NRAS', 'KRAS'))
@

\paragraph{ Here we try to include also a soft exclusivity (OR) pattern but, since its \textit{"signature"}
is the same of the hard one, it will not be included. The code below is commented because it gives errors. }
<<eval=FALSE>>=
hypo = hypothesis.add(hypo, 'NRAS or KRAS',  OR('NRAS', 'KRAS'))
@

\paragraph{ For the sake to better highlight the perfect (hard) exclusivity between NRAS/KRAS 
mutations one can visualize their alterations }
<<>>=
dummy = oncoprint(events.selection(hypo, filter.in.names = c('KRAS', 'NRAS')))
@
<<include=FALSE,echo=FALSE>>=
capture.output(oncoprint(events.selection(hypo, filter.in.names = c('KRAS', 'NRAS')), file='onco-3.pdf'), file='NUL')
@
\incfig[ht]{onco-3}{0.9\textwidth}{Oncoprint output}{}

\FloatBarrier

\paragraph{ This is as above, but includes other events. Again, we can include only the hard exclusivity pattern }
<<>>=
hypo = hypothesis.add(hypo, 'SF3B1 xor ASXL1', XOR('SF3B1', OR('ASXL1')), '*')
@
<<eval=FALSE>>=
hypo = hypothesis.add(hypo, 'SF3B1 or ASXL1', OR('SF3B1', OR('ASXL1')), '*')
@

\paragraph{ We now do the same for TET2 and IDH2. In this case 3 events for TET2 are present, which are
\textit{"Ins/Del"}, \textit{"Missense point"} and \textit{"Nonsense point"}. For this reason, since we are not specifying
a subset of such events all TET2 alterations are used. Since these show a perfect hard exclusivity
trend these will be included in XOR. }
<<>>=
as.events(hypo, genes = 'TET2') 
hypo = hypothesis.add(hypo, 'TET2 xor IDH2', XOR('TET2', 'IDH2'), '*')
@
<<eval=FALSE>>=
hypo = hypothesis.add(hypo,  'TET2 or IDH2', OR('TET2', 'IDH2'), '*')
@

<<>>=
dummy = oncoprint(events.selection(hypo, filter.in.names = c('TET2', 'IDH2')))
@
<<include=FALSE,echo=FALSE>>=
capture.output(oncoprint(events.selection(hypo, filter.in.names = c('TET2', 'IDH2')), file='onco-4.pdf'), file='NUL')
@
\incfig[ht]{onco-4}{0.9\textwidth}{Oncoprint output}{}

\FloatBarrier

\paragraph{ For every gene that has more than one event associated we also add a soft exclusivity pattern
for its events }
<<>>=
hypo = hypothesis.add.homologous(hypo)
@

\paragraph{ The dataset input to CAPRI is shown }
<<>>=
dummy = oncoprint(hypo,  gene.annot = list(priors= gene.hypotheses), sample.id = T)
@
<<include=FALSE,echo=FALSE>>=
capture.output(oncoprint(hypo,  gene.annot = list(priors= gene.hypotheses), sample.id = T, file='onco-5.pdf'), file='NUL')
@
\incfig[ht]{onco-5}{0.9\textwidth}{Oncoprint output}{}

\FloatBarrier

\paragraph{\large Model reconstruction}{\ }\\

\paragraph{ We execute CAPRI with its default parameter: we use both AIC/BIC regularizators, Hill-climbing
exhaustive bootstrap (100 replicates for Wilcoxon testing), p-value 0.05 and we set seed }
<<>>=
model = tronco.capri(hypo, boot.seed = 12345, regularization='bic', nboot=6)
@

\paragraph{ We can plot the reconstructed model. We set some parameters to get a fancy plot; confidence
is shown as temporal priority and probability raising (selective advantage scores) and 
hypergeometric testing (goodness of input data). }
<<figplot, fig=TRUE,include=FALSE>>=
tronco.plot(model,  
  fontsize = 13,  
  scale.nodes = .6,   
  confidence = c('tp', 'pr', 'hg'),   
  height.logic = 0.25,  
  legend.cex = .5, 
  pathways =  list(priors= gene.hypotheses))
@
\incfig[ht]{vignette-figplot}{0.9\textwidth}{aCML Reconstructed model}
{Pre bootstrap.}

\FloatBarrier

\paragraph{\large Bootstrapping data}{\ }\\

<<>>=
model.boot = tronco.bootstrap(model, nboot=6)
@

<<figplotboot, fig=TRUE,include=FALSE>>=
tronco.plot(model.boot,
            fontsize = 13,  
            scale.nodes = .6,
            confidence=c('npb'),
            height.logic = 0.25,  
            legend.cex = .5)
@
\incfig[ht]{vignette-figplotboot}{0.9\textwidth}{aCML Reconstructed model}
{After bootstrap.}


\end{document}



% buggone -> fix http://stackoverflow.com/questions/12481267/in-r-how-to-prevent-blank-page-in-pdf-when-using-gridbase-to-embed-subplot-insi ??

